### PR TITLE
feat(OpenGLRendering): Add radon blending mode

### PR DESCRIPTION
**Pull request by laurennlam**
### Context
Add the possibility to use the radon transform in blending mode.

### Results
image

### Changes
Updated vtkVolumeFS glsl file to introduce radon transform.

### PR and Code Checklist
- [x] Clean example code to fit with previous example and adapt values to the loaded volume

### Testing
You can see the changes by running the VolumeMapperBlendModes example.

@finetjul @martinken 
The first needed code review is for the glsl file. I'll update the example code later.

### CODE REVIEW

File: Examples/Volume/VolumeMapperBlendModes/index.js [hunk 1]
```diff
@@ -50,13 +53,33 @@ ofun.addPoint(-637.62, 0.1);
...
-const ctfun = vtkColorTransferFunction.newInstance();
-ctfun.addRGBPoint(-3024, 0, 0, 0);
-ctfun.addRGBPoint(-637.62, 0, 0, 0);
-ctfun.addRGBPoint(700, 1, 1, 1);
-ctfun.addRGBPoint(3071, 1, 1, 1);
+let min = 200;
+let max = 760;
+let maxValue = 0.17;
+let minValue = 0;
+let unit = 0.3;
```
Comment (finetjul) [inline Examples/Volume/VolumeMapperBlendModes/index.js:60]: please set those into a unique object:
```
const params  = {
 min: 200,
 max: 760,
 ...
};
```

File: Examples/Volume/VolumeMapperBlendModes/index.js [hunk 2]
```diff
@@ -68,56 +91,61 @@ actor.getProperty().setSpecularPower(10.0);
...
+}
 
+function updateScalarMax(event) {
+  max = event.target.value;
+  updateColorFunction();
   renderWindow.render();
 }
 
-function updateScalarMin(event) {
-  mapper.setIpScalarRange(event.target.value, mapper.getIpScalarRange()[1]);
```
Comment (finetjul) [inline Examples/Volume/VolumeMapperBlendModes/index.js:97]: if you no longer set those, is this still useful ?

File: Examples/Volume/VolumeMapperBlendModes/index.js [hunk 3]
```diff
@@ -68,56 +91,61 @@ actor.getProperty().setSpecularPower(10.0);
...
-  } else {
-    // Hide scalar ui
-    for (let i = 0; i < ipScalarEls.length; i += 1) {
-      const el = ipScalarEls[i];
-      el.style.display = 'none';
-    }
-  }
+function updateScalarMin(event) {
+  min = event.target.value;
+  updateColorFunction();
```
Comment (daker) [inline Examples/Volume/VolumeMapperBlendModes/index.js:96]: maybe this shoube event.target.valueAsNumber

File: Sources/Rendering/OpenGL/glsl/vtkVolumeFS.glsl [hunk 4]
```diff
@@ -1330,6 +1330,39 @@ void applyBlend(vec3 posIS, vec3 endIS, float sampleDistanceIS, vec3 tdims)
...
+      // now map through opacity and color
+      tColor = getColorForValue(tValue, posIS, tstep);
+
+      color = color - vec4(tColor.rgb*sampleDistance*tColor.a, 0);
+      posIS += stepIS;
+      stepsTraveled++;
+    }
+
+    gl_FragData[0] = vec4(color.rgb, color.a);
+  #endif
```
Comment (finetjul) [inline Sources/Rendering/OpenGL/glsl/vtkVolumeFS.glsl:1365]: @martinken just in case you got time to review this last one... 🙏

**Comment by laurennlam** [comment]
@finetjul Ready for review

File: Examples/Volume/VolumeMapperBlendModes/index.js [hunk 5]
```diff
@@ -31,10 +31,12 @@ fullScreenRenderer.addController(controlPanel);
 // ----------------------------------------------------------------------------
 // Example code
 // ----------------------------------------------------------------------------
-// Server is not sending the .gz and with the compress header
-// Need to fetch the true file name and uncompress it locally
-// ----------------------------------------------------------------------------
-
+let currentBlendMode = null;
```
Comment (finetjul) [inline Examples/Volume/VolumeMapperBlendModes/index.js:34]: why bother exposing this variable when you can replace it with `mapper.getBlendMode()`

File: Examples/Volume/VolumeMapperBlendModes/index.js [hunk 6]
```diff
@@ -90,16 +180,22 @@ function updateBlendMode(event) {
...
+      const el = radonScalars[i];
+      el.style.display = 'table-row';
+    }
+  } else {
+    // Hide scalar ui
+    for (let i = 0; i < radonScalars.length; i += 1) {
+      const el = radonScalars[i];
+      el.style.display = 'none';
+    }
+  }
```
Comment (finetjul) [inline Examples/Volume/VolumeMapperBlendModes/index.js:196]: Factorize as such ?
``` 
for (let i = 0; i < radonScalars.length; i += 1) {
      const el = radonScalars[i];
      el.style.display = currentBlendMode === 5 ? 'table-row : 'none';
```

File: Examples/Volume/VolumeMapperBlendModes/controller.html [hunk 7]
```diff
@@ -8,15 +8,40 @@
...
+  <tr class="radonScalar"  style="display:none;">
+    <td>First low hounsfield :</td>
+    <td><input class="min" type="range" min="-1000" max="2500" step="5"></td>
+    <td><label id="minLabel"></label></td>
+    <td>Value :</td>
+    <td><input class="minValue" type="range" min="0" max="0.2" step="0.001"></td>
+    <td><label id="minValueLabel"></label></td>
+  </tr>
+  <tr class="radonScalar"  style="display:none;">
+    <td>First max hounsfield:</td>
```
Comment (finetjul) [inline Examples/Volume/VolumeMapperBlendModes/controller.html:35]: low -> high
min -> max
but not a mix of both

File: Sources/Rendering/OpenGL/glsl/vtkVolumeFS.glsl [hunk 8]
```diff
@@ -1330,6 +1330,39 @@ void applyBlend(vec3 posIS, vec3 endIS, float sampleDistanceIS, vec3 tdims)
 
     gl_FragData[0] = getColorForValue(sum, posIS, tstep);
   #endif
+  #if vtkBlendMode == 5 // RADON
```
Comment (finetjul) [inline Sources/Rendering/OpenGL/glsl/vtkVolumeFS.glsl:1333]: Please update API documentation to inform about this new blend mode

File: Examples/Volume/VolumeMapperBlendModes/index.js [hunk 9]
```diff
@@ -50,13 +52,54 @@ ofun.addPoint(-637.62, 0.1);
...
+      radonParameters.maxValue,
+      radonParameters.maxValue,
+      radonParameters.maxValue
+    );
+    ctfun.addRGBPoint(
+      3071,
+      radonParameters.maxValue,
+      radonParameters.maxValue,
+      radonParameters.maxValue
+    );
```
Comment (finetjul) [inline Examples/Volume/VolumeMapperBlendModes/index.js:85]: please move the CTF creation into a convenient helper function (in the class that has setBlendMode)

Comment (finetjul) [reply Examples/Volume/VolumeMapperBlendModes/index.js:85]: I'm also surprised that those radonParameters have no impact on `ofun`

File: Examples/Volume/VolumeMapperBlendModes/index.js [hunk 10]
```diff
@@ -50,13 +52,54 @@ ofun.addPoint(-637.62, 0.1);
...
+
+    minLabel.innerHTML = radonParameters.min;
+    maxLabel.innerHTML = radonParameters.max;
+    maxValueLabel.innerHTML = radonParameters.maxValue;
+    minValueLabel.innerHTML = radonParameters.minValue;
+    unitLabel.innerHTML = radonParameters.unit;
+    mapper.setSampleDistance(radonParameters.unit);
```
Comment (finetjul) [inline Examples/Volume/VolumeMapperBlendModes/index.js:92]: `unit` does not seem proper to radon. Move it outside radon ?

Comment (laurennlam) [reply Examples/Volume/VolumeMapperBlendModes/index.js:92]: This is not proper to radon, but it wasn't exposed on the other blend type in the initial example. I don't want to change the behavior of it. So I will reinitialize sample distance to initial value when setting blend mode other than radon.

File: Sources/Rendering/Core/VolumeMapper/Constants.js [hunk 11]
```diff
@@ -4,6 +4,7 @@ export const BlendMode = {
   MINIMUM_INTENSITY_BLEND: 2,
   AVERAGE_INTENSITY_BLEND: 3,
   ADDITIVE_INTENSITY_BLEND: 4,
+  RADON: 5,
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/Constants.js:7]: RADON_BLEND

File: Sources/Rendering/Core/VolumeMapper/Constants.d.ts [hunk 12]
```diff
@@ -4,8 +4,9 @@ export declare enum BlendMode {
 	MINIMUM_INTENSITY_BLEND = 2,
 	AVERAGE_INTENSITY_BLEND = 3,
 	ADDITIVE_INTENSITY_BLEND = 4,
+	RADON = 5,
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/Constants.d.ts:7]: RADON_BLEND

File: Sources/Rendering/Core/VolumeMapper/index.d.ts [hunk 13]
```diff
@@ -133,6 +133,11 @@ export interface vtkVolumeMapper extends vtkAbstractMapper {
 	 */
 	setBlendModeToAverageIntensity(): void;
 
+	/**
+	 * Set blend mode to RADON
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.d.ts:137]: Set Blend mode to RADON transform (integral)

File: Sources/Rendering/Core/VolumeMapper/Constants.js (hunk 11 above)
Comment (finetjul) [reply Sources/Rendering/Core/VolumeMapper/Constants.js:7]: or maybe RADON_TRANSFORM_BLEND ?

File: Examples/Volume/VolumeMapperBlendModes/index.js (hunk 9 above)
Comment (finetjul) [reply Examples/Volume/VolumeMapperBlendModes/index.js:86]: Why is it resolved but I do not see a convenient helper function ?

Comment (laurennlam) [reply Examples/Volume/VolumeMapperBlendModes/index.js:86]: I added a helper function : createCTF()

Comment (finetjul) [reply Examples/Volume/VolumeMapperBlendModes/index.js:86]: my bad, I meant to add a convenient `createRadonTransformTransferFunctions(...)` static helper function in `VolumeMapper` that returns both a CTF and PiecewiseFunction

File: Examples/Volume/VolumeMapperBlendModes/index.js (hunk 10 above)
Comment (finetjul) [reply Examples/Volume/VolumeMapperBlendModes/index.js:87]: Let's expose it to all blend modes to see the impact of the sample distance on each blend mode

File: Sources/Rendering/Core/VolumeMapper/index.d.ts [hunk 14]
```diff
@@ -233,6 +238,19 @@ export interface vtkVolumeMapper extends vtkAbstractMapper {
...
+/**
+ * create the piecewise and the color transfer functions for the radon blending.
+ * @static
+ * @param {object} values: Store all needed values to create the CTF:
+ * {
+ *   min: Define minimum voxel value
+ *   minValue: Value that will be attribute to voxels with min value
+ *   max: Define maximum voxel value
+ *   maxValue: Value that will be attribute to voxels with max value
+ * }
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.d.ts:250]: Please provide more context and help to choose the parameter values.
(e.g. to be used with the RADON_TRANSFORM blend mode, hounsfield units, CT volume...)
Talk about absorption...
Why in your example minValue is > than maxValue ?

File: Sources/Rendering/Core/VolumeMapper/index.d.ts [hunk 15]
```diff
@@ -233,6 +238,19 @@ export interface vtkVolumeMapper extends vtkAbstractMapper {
 	update(): void;
 }
 
+/**
+ * create the piecewise and the color transfer functions for the radon blending.
+ * @static
+ * @param {object} values: Store all needed values to create the CTF:
+ * {
+ *   min: Define minimum voxel value
+ *   minValue: Value that will be attribute to voxels with min value
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.d.ts:247]: will be attribute**d**

File: Sources/Rendering/Core/VolumeMapper/index.d.ts [hunk 16]
```diff
@@ -233,6 +238,19 @@ export interface vtkVolumeMapper extends vtkAbstractMapper {
 	update(): void;
 }
 
+/**
+ * create the piecewise and the color transfer functions for the radon blending.
+ * @static
+ * @param {object} values: Store all needed values to create the CTF:
+ * {
+ *   min: Define minimum voxel value
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.d.ts:246]: what does it mean minimum voxel value. Is it the same as `getPointsData().getScalars().getRange()[0]`?
Is it in hounsfield unit ?

File: Sources/Rendering/Core/VolumeMapper/index.d.ts [hunk 17]
```diff
@@ -233,6 +238,19 @@ export interface vtkVolumeMapper extends vtkAbstractMapper {
 	update(): void;
 }
 
+/**
+ * create the piecewise and the color transfer functions for the radon blending.
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.d.ts:242]: piecewise -> opacity

File: Sources/Rendering/Core/VolumeMapper/index.js [hunk 18]
```diff
@@ -2,9 +2,45 @@ import macro from 'vtk.js/Sources/macros';
...
+import vtkColorTransferFunction from 'vtk.js/Sources/Rendering/Core/ColorTransferFunction';
 
 const { BlendMode, FilterMode } = Constants;
 
+function createRadonTransferFunctions(values) {
+  const ofun = vtkPiecewiseFunction.newInstance();
+  ofun.addPoint(-3024, 0.1);
+  ofun.addPoint(-637.62, 0.1);
+  ofun.addPoint(700, 0.5);
+  ofun.addPoint(3071, 0.9);
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.js:15]: why is it hard-coded and does not depend on `values` ?

File: Sources/Rendering/Core/VolumeMapper/index.js [hunk 19]
```diff
@@ -2,9 +2,45 @@ import macro from 'vtk.js/Sources/macros';
 import * as vtkMath from 'vtk.js/Sources/Common/Core/Math';
 import Constants from 'vtk.js/Sources/Rendering/Core/VolumeMapper/Constants';
 import vtkAbstractMapper from 'vtk.js/Sources/Rendering/Core/AbstractMapper';
+import vtkPiecewiseFunction from 'vtk.js/Sources/Common/DataModel/PiecewiseFunction';
+import vtkColorTransferFunction from 'vtk.js/Sources/Rendering/Core/ColorTransferFunction';
 
 const { BlendMode, FilterMode } = Constants;
 
+function createRadonTransferFunctions(values) {
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.js:10]: `values` should not be an `object`, but you must create multiple parameters

File: Sources/Rendering/Core/VolumeMapper/index.js [hunk 20]
```diff
@@ -2,9 +2,45 @@ import macro from 'vtk.js/Sources/macros';
...
+  const ofun = vtkPiecewiseFunction.newInstance();
+  ofun.addPoint(-3024, 0.1);
+  ofun.addPoint(-637.62, 0.1);
+  ofun.addPoint(700, 0.5);
+  ofun.addPoint(3071, 0.9);
+
+  const ctfun = vtkColorTransferFunction.newInstance();
+  ctfun.addRGBPoint(-3024, 0.0, 0.0, 0.0);
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.js:18]: why -3024 and not -1024 ?

File: Sources/Rendering/Core/VolumeMapper/index.js [hunk 21]
```diff
@@ -2,9 +2,45 @@ import macro from 'vtk.js/Sources/macros';
...
+    values.minValue,
+    values.minValue
+  );
+  ctfun.addRGBPoint(
+    values.max,
+    values.maxValue,
+    values.maxValue,
+    values.maxValue
+  );
+  ctfun.addRGBPoint(3071, values.maxValue, values.maxValue, values.maxValue);
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.js:31]: Here in red is the look of your grayscale color transfer function (where `M1 = min(values.min, values.max)` and `M2 = max(values.min, values.max)`) :
image

Don't you want the blue line instead ?

File: Sources/Rendering/Core/VolumeMapper/index.js [hunk 22]
```diff
@@ -2,9 +2,45 @@ import macro from 'vtk.js/Sources/macros';
...
+
+  return { piecewiseFunction: ofun, colorTransferFunction: ctfun };
```
Comment (finetjul) [inline Sources/Rendering/Core/VolumeMapper/index.js:33]: piecewiseFunction -> opacityFunction

**Review by finetjul — CHANGES_REQUESTED** [review]

File: Sources/Rendering/Core/VolumeMapper/index.js (hunk 18 above)
Comment (laurennlam) [reply Sources/Rendering/Core/VolumeMapper/index.js:15]: These values already been set before. So we let it as it was.

File: Sources/Rendering/Core/VolumeMapper/index.js (hunk 20 above)
Comment (laurennlam) [reply Sources/Rendering/Core/VolumeMapper/index.js:18]: In order to fit with the piecewise function.

File: Sources/Rendering/Core/VolumeMapper/index.d.ts (hunk 14 above)
Comment (laurennlam) [reply Sources/Rendering/Core/VolumeMapper/index.d.ts]: This is an error

File: Sources/Rendering/Core/VolumeMapper/index.d.ts (hunk 16 above)
Comment (laurennlam) [reply Sources/Rendering/Core/VolumeMapper/index.d.ts]: Description of the method has been improved to explain it.

File: Sources/Rendering/Core/VolumeMapper/index.js (hunk 21 above)
Comment (laurennlam) [reply Sources/Rendering/Core/VolumeMapper/index.js]: After testing the blue line, it doesn't seem to have any difference. But if you prefer, I can setup the blue line.

File: Sources/Rendering/OpenGL/glsl/vtkVolumeFS.glsl [hunk 23]
```diff
@@ -1435,6 +1435,39 @@ void applyBlend(vec3 posIS, vec3 endIS, vec3 tdims)
...
+
+      // now map through opacity and color
+      tColor = getColorForValue(tValue, posIS, tstep);
+
+      color = color - vec4(tColor.rgb*sampleDistance*tColor.a, 0);
+      posIS += stepIS;
+      stepsTraveled++;
+    }
+
+    gl_FragData[0] = vec4(color.rgb, color.a);
```
Comment (finetjul) [inline Sources/Rendering/OpenGL/glsl/vtkVolumeFS.glsl:1469]:
```
color = 1
for (int i = 0; i < ... )
{
  tValue = getTextureValue(posIS);
  tColor = getColorForValue(posIS, posIS, tstep);
  color = color - tColor.a * sampleDistance;
  posIS += stepIS;
  stepsTraveled++;
} 
gl_FragData[0] = texture2D(ctexture, vec2(color * cscale0 + cshift0, 0.5));
```

**Comment by laurennlam** [comment]
@finetjul Ready for review/merge

**Pull request merged**